certificateshop-logic/src/main/java/com/epam/esm/service/impl/UserServiceImpl.java

```diff
@@ -0,0 +1,269 @@
+        List<String> errorMessage = new ArrayList<>();
+        int pageNumber = Integer.parseInt(parameters.get(ColumnNames.PAGE_NUMBER_PARAM_NAME));
+        int amountEntitiesOnThePage = Integer.parseInt(parameters.get(ColumnNames.AMOUNT_OF_ENTITIES_ON_THE_PAGE_PARAM_NAME));
+        checkLimitAndOffset(errorMessage, pageNumber, amountEntitiesOnThePage);
+        List<User> users = userDao.findAllPagination(pageNumber, amountEntitiesOnThePage);
+        return users;
+    }
+
+    private void checkLimitAndOffset(List<String> errorMessage, long offset, long limit) {
+        if (offset < 0) {
+            errorMessage.add(translator.toLocale("THE_OFFSET_SHOULD_BE_MORE_THAN_0"));
+        }
+        if (limit < 0) {
+            errorMessage.add(translator.toLocale("THE_LIMIT_SHOULD_BE_MORE_THAN_0"));
+        }
+        if (!errorMessage.isEmpty()) {
+            throw new MethodArgumentNotValidException(
+                    ERROR_CODE_METHOD_ARGUMENT_NOT_VALID + ERROR_CODE_USER_NOT_VALID, errorMessage);
+        }
+    }
+
+    /**
+     * Creates a new {@link User} in the system.
+     *
+     * @param user is the {@link User} to create.
+     */
+    @Override
+    public User createUser(User user) {
+        List<String> errorMessage = new ArrayList<>();
+        if (user.getNickName() != null && !user.getNickName().trim().equals("")) {
+            checkIfUserExistInSystem(user, errorMessage);
+            if (user.getOrders() == null || user.getOrders().isEmpty()) {
+                return createNewUser(user);
+            } else {
+                for (Order order : user.getOrders()) {
+                    if (order.getCertificates() == null || order.getCertificates().isEmpty()) {
+                        errorMessage.add(translator.toLocale("CERTIFICATES_IN_ORDER_SHOULD_BE_NOT_EMPTY"));
+                        throw new MethodArgumentNotValidException(
+                                ERROR_CODE_ENTITY_NOT_FOUND + ERROR_CODE_USER_NOT_VALID, errorMessage);
+                    }
+                    for (GiftCertificate certificate : order.getCertificates()) {
+                        checkCertificateId(errorMessage, certificate);
+                    }
+                }
+                if (!errorMessage.isEmpty()) {
+                    throw new MethodArgumentNotValidException(
+                            ERROR_CODE_ENTITY_NOT_FOUND + ERROR_CODE_USER_NOT_VALID, errorMessage);
+                }
+                User createdUser = createNewUser(user);
+                for (Order order : user.getOrders()) {
+                    addOrdersToUser(user, createdUser, order, errorMessage);
+                }
+            }
+        } else {
+            errorMessage.add(translator.toLocale("THE_FIELD_NICKNAME_SHOULD_BE_NOT_EMPTY"));
+            throw new MethodArgumentNotValidException(ERROR_CODE_DUPLICATE + ERROR_CODE_USER_NOT_VALID, errorMessage);
+        }
+        return userDao.findByName(user.getNickName()).get();
```

Comment:
Do you have a test for this method? If so, do you created tests for ALL cases?

Reply:
Divide this method into many small private methods because of IFs

Reply:
done. I have not written tests in this module. But if it is mandatory i will.

Reply:
UserServiceTest and OrderServiceTest are written